### ISSUE TITLE
rules for ELF/Linux sample

### ISSUE THREAD

**Issue by williballenthin** [mandiant/capa-rules#736]
even when the OS is specified manually, we don't get many results. we should spend some time writing rules that match the interesting behavior for this ELF/Linux sample:

```
capa tests/data/72f1b91327ffda4cf18a2bf64913b673d39ebbff8cbe50c9cd354b1dcd312bcc --os linux
matching: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 297/297 [00:04<00:00, 63.43 functions/s, skipped 0 library functions]
+------------------------+------------------------------------------------------------------------------------+
| md5                    | 2bf18d0403677378adad9001b1243211                                                   |
| sha1                   | 21693bf5c29c5dbc973047e0c1607ebdd000de9b                                           |
| sha256                 | 72f1b91327ffda4cf18a2bf64913b673d39ebbff8cbe50c9cd354b1dcd312bcc                   |
| os                     | linux                                                                              |
| format                 | elf                                                                                |
| arch                   | amd64                                                                              |
| path                   | tests/data/72f1b91327ffda4cf18a2bf64913b673d39ebbff8cbe50c9cd354b1dcd312bcc        |
+------------------------+------------------------------------------------------------------------------------+



+------------------------------------------------------+------------------------------------------------------+
| CAPABILITY                                           | NAMESPACE                                            |
|------------------------------------------------------+------------------------------------------------------|
| reference Google Public DNS server                   | communication/dns                                    |
+------------------------------------------------------+------------------------------------------------------+
```

_Originally posted by @williballenthin in https://github.com/mandiant/capa/issues/1402#issuecomment-1484040810_

**Comment by williballenthin** [comment]
<img width="391" alt="image" src="https://user-images.githubusercontent.com/156560/227766540-893a13da-2c68-4460-843b-2f1540517475.png">

<img width="338" alt="image" src="https://user-images.githubusercontent.com/156560/227766544-23c287c5-0db9-472b-81a8-a7581a0bca6e.png">

<img width="178" alt="image" src="https://user-images.githubusercontent.com/156560/227766563-cf2e7fff-d8a1-41b5-8bf6-160b97a40688.png">

<img width="220" alt="image" src="https://user-images.githubusercontent.com/156560/227766594-b8e28ae7-9d7c-4416-b4af-190f899efc5d.png">

<img width="325" alt="image" src="https://user-images.githubusercontent.com/156560/227766606-605a78d1-bb95-4e7c-bd21-26ddeb8b464c.png">

<img width="195" alt="image" src="https://user-images.githubusercontent.com/156560/227766642-969e1406-ea15-4dc2-bd76-ad7a86e3de4a.png">

**Comment by williballenthin** [comment]
i wonder if maybe viv is not correctly resolving the symbol names for these routines, like `socket` and `recv`

**Comment by williballenthin** [comment]
of course, you can always open PRs to the capa-rules repository with proposed new rules. feel free to identify any interesting behavior in this sample and figure out how to create rules that match it. if there are bugs with extracting or matching features, please update this issue with your findings.

**Comment by yelhamer** [comment]
The sample in question is statically linking some of libc's syscall wrapper functions, which are implemented using the _syscall_ instruction. As a result, viv cannot resolve the symbol names for functions such as _send_, _socket_, _recv_, etc. (becasue they have been statically linked), and VivFeatureExtractor can't either because it's looking for call/jmp instructions when instead it should be looking for _syscall_ instructions.

The potential fixes I propose are the following:
- **Adjust the ruleset:** use the mnemonic and number feature to find specific syscalls. Example:
```yaml
...
- api: ws2_32.send
- api: send
- and:
  - mnemonic: syscall
  - number: 0x2c = send()
... 
```
- **Add support for extracting api's from syscall:** I wonder if it would be possible to add logic to VivFeatureExtractor so that whenever extract_insn_api_features() encounters a _syscall_ it goes back one instruction, tries to fetch the syscall number, and looks it up in a os-specific dictionary search to find the syscall being referenced; however, this approach assumes that the instruction **mov (r|e)ax, call_num** would always be right behind the _syscall_/_int 0x80_ instruction.
- **Add extra logic for fetching api's from statically linked binaries:** In this particular samples, most imported functions tend to have the format: <__libc_function-name>. I wonder if this generalized across all versions of libc, and whether it would be worth it to add logic to capa that looks specifically for these names and tries to parse them if it detects that the sample has been statically linked.

What are your thoughts on this @williballenthin ? should we modify existing rules that rely on libc's syscall wrapper functions according to the first suggestion, or should we explore the second and/or third options?

**Comment by yelhamer** [comment]
one issue with the first approach is that rule authors would need to account for different operating systems having different symbol tables, which might make rules more needlessly complex, and might also lead to false reports if rule authors forget to specify which _os_ the syscall number is meant for.

for example if someone is using custom rules that were intended for linux to scan a *bsd sample, capa might end up mischaracterizing a possible syscall 41 as a socket() when in this context it's a dup(), thereby leading to a false capability detection (create TCP socket)

**Comment by williballenthin** [comment]
(I’ve been thinking about this since Friday and will have time shortly to give a good response. Thanks for the deep thought about this issue and practical ideas for consideration)

**Comment by williballenthin** [comment]
I think we should use a multi-pronged strategy to address two aspects of this issue:

  1. we should recognize raw syscalls for what they are, and turn them into something human readable, and also
  2. we should recognize known syscall wrappers so that we can reason about them.

(2) is important because capa can't really relate logic found in one function with the logic of another function. @mr-tz and I have debated this for hours and can't figure out how to make it work intuitively (so, let us know if you have a clever idea!).

<details>
<summary>here's why</summary>

capa might look at this function:

![image](https://user-images.githubusercontent.com/156560/230978134-c73d5c84-2bf7-47aa-a7bf-7bbaa0929257.png)

and recognize "this connects a socket"

and independently, capa might look at this function:

![image](https://user-images.githubusercontent.com/156560/230978381-e97c7064-8ef4-4256-822a-753e57dc888a.png)

and recognize "there are two calls and some conditionals" (it doesn't know the names of the functions, there's no debug info)

and it *cannot* use its conclusions about `connect_wrapper` when it analyzes the caller function. each function scope is evaluated independently.

</details>

so, i think we should:

  a. create a collection of rules to identify behaviors by raw syscall, like "create a file via syscall" that might look like:
```yml
- basic block:
  - and:
    - os: linux  # note we can restrict by OS so to avoid confusion with *BSD, etc.
    - mnemonic: syscall
    - instruction:
      - mnemonic: mov
      - operand[1].number: 2 = sys_open
      # unfortunately cannot restrict to rax today
```

  b. develop better function ID technology to identify the syscall wrappers for what they are and let other rules reference them as API features, like `api: open`. some options:
    i. ensure debug info is used, when present, and
    ii. develop and distribute FLIRT signatures for common GCC wrapper functions, or
    iii. develop a custom analysis pass for vivisect to identify GCC wrappers manually


(a) ensures we can at least conclude "this sample can open a file (via syscall)" while (b) lets us reason about things like "the sample can open a file and also it references the path '/etc/passwd' so it steals passwords" (or whatever).

i suppose (a) also helps if the compiler inlines the syscall wrapper or if the author uses syscalls directly to be tricky.

**Comment by williballenthin** [comment]
for (a)(i) and the given sample, as you note, the connect wrapper function:

![image](https://user-images.githubusercontent.com/156560/230982353-99d37f17-7c81-48b2-9d72-9fe109a62850.png)

has ELF metadata that indicates the function name:

![image](https://user-images.githubusercontent.com/156560/230982521-4a6dc2d7-3175-4fa9-91ab-43edac072579.png)

which we *definitely* want capa to use for `api` features

**Comment by williballenthin** [comment]
for (b)(ii/iii), for the connect wrapper, we have:

![image](https://user-images.githubusercontent.com/156560/230985095-db85e332-12f6-4e96-87b6-8d833d75eea3.png)

which have the relevant byte sequence:

```
53 89 D2 48 63 FF B8 2A 00 00 00 0F 05 48 3D 00 F0 FF FF 48 89 C3
                     ^^ syscall number
```

how stable is this across GCC/libc versions? 

some rough numbers:

![image](https://user-images.githubusercontent.com/156560/230986282-e256e21e-68e8-4f74-bd2d-cc29e73d1896.png)


![image](https://user-images.githubusercontent.com/156560/230986317-5f0b1bf6-5f19-4d59-aaed-37c3bc839639.png)

**Comment by yelhamer** [comment]
I really like this approach @williballenthin !

I have some general comments outlined bellow:

>(2) is important because capa can't really relate logic found in one function with the logic of another function. @mr-tz and I have debated this for hours and can't figure out how to make it work intuitively (so, let us know if you have a clever idea!).

(a) I feel like this should be really useful/important in this particular case, since if a static binary gets stripped then we would no longer be able to use approach (2)(i) for that sample; however, (2)(ii/iii) should still work.

(b) Regarding (1), I think that we need to some testing after we write the rules; since I believe we might get more false positives that are due to the feature extractor mislabeling syscall params as syscall numbers. For example, if a syscall has an imm parameter of 2, then shouldn't it trigger the "opens a file using sys_open()" capability regardless of the nature of the syscall being made?

The priorities you suggested look good to me. I will take a deeper look at this — especially approach (2) — tomorrow, and I will try to write come code/rules for this.

Thank you so much for the elaborate thoughts and explanation!

**Comment by xusheng6** [comment]
As a note, BN does a decent job in recognizing the system calls and convert them into a human readable name. We do so in the BN `platform`, which has a system call id-name mapping in it. If vivisetc wishes do something similar, this might be a viable approach.
